modules/mauth-authenticator-apachehttp/src/main/java/com/mdsol/mauth/apache/HttpClientPublicKeyProvider.java

```diff
@@ -102,9 +116,7 @@ private class PublicKeyResponseHandler implements ResponseHandler<String> {
     @Override
     public String handleResponse(HttpResponse response) throws IOException {
       if (response.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
-        if (ttl == null) {
-          ttl = getMaxAge(response).orElse(configuration.getTimeToLive());
-        }
+        ttl = getMaxAge(response).orElse(configuration.getTimeToLive());
```

Comment:
I think we probably want to fix more things in this class. 

ttl should not be a field that we modify on every request sent - this class is going to be used in a multithreaded context so it's just not right..

Here's what I suggest we do:

* Make PublicKeyResponseHandler return a String (the pub key) and a Long (the age)
* Make the cache store both the pub key and the max age, you can then use [Expiry](https://github.com/ben-manes/caffeine/wiki/Eviction#time-based) to calculate the expiry time (since max age is now part of the cache value)
* Remove the whole mutex thing. It's going to lead to massive performance problems as only one caller can call `getPublicKey` at a time...

Reply:
@jatcwang thanks a lot. Updated the cache as suggested in https://github.com/mdsol/mauth-jvm-clients/pull/142/commits/4fded8e97d7b44f386b8f81b1bf50eda53fbe254